cloudinit/sources/DataSourceEc2.py

```diff
@@ -496,20 +523,24 @@ def _refresh_stale_aws_token_cb(self, msg, exception):
             self._api_token = None
         return True  # always retry
 
-    def _status_cb(self, msg, exc=None):
-        LOG.warning(msg)
-        if 'Read timed out' in msg:
-            LOG.warning('Cannot use Ec2 IMDSv2 API tokens, using IMDSv1')
-            self._api_token = API_TOKEN_DISABLED
+    def _imds_exception_cb(self, msg, exception=None):
+        if isinstance(exception, uhelp.UrlError) and exception.code >= 400:
+            if exception.code == 403:
+                LOG.warning('Ec2 IMDS endpoint returned a 403 error. '
+                            'HTTP endpoint is disabled. Aborting.')
+            else:
+                LOG.warning('Fatal error while requesting '
+                            'Ec2 IMDSv2 API tokens')
+            raise exception
```

Comment:
This is interesting data.  Can you get me instance IDs where you have seen this?  It doesn't match the guidance I have received  from folks maintaining IMDS. I would like to dive more into this.   I can modify the code to keep retrying on 400s for now and revisit later. Would that work for you?

Reply:
@fred-lefebvre there error was on my part, well cloud-init's part in Focal only. 
Here's a fix that we will get into Ubuntu Focal which points to the details 
https://github.com/canonical/cloud-init/pull/230

LP:1865882

Reply:
> But I believe that was your desired outcome here right?

Yes. I'm removing the DataSourceNone from the list of data sources we use in our build (for now) to avoid the remapping of the instance_link to the fake instance id provided by DataSourceNone. This is actually the problem I'm trying to address here : https://github.com/canonical/cloud-init/pull/229